### PR TITLE
feat(hs): chunking data based on batch limit

### PR DESCRIPTION
**Pull request by mihir-4116**
## What are the changes introduced in this PR?

Write a brief explainer on your code changes.

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Type of change

If the pull request is a **bug-fix**, **enhancement** or a **refactor**, please fill in the details on the changes made.

- Existing capabilities/behavior

- New capabilities/behavior

If the pull request is a **new feature**,

### Any technical or performance related pointers to consider with the change?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### If the PR has changes in more than 10 files, please mention why the changes were not split into multiple PRs.

N/A

### If multiple linear tasks are associated with the PR changes, please elaborate on the reason:

N/A

<hr>

### Developer checklist

- [ ] **No breaking changes are being introduced.**

- [ ] Are all related docs linked with the PR?

- [ ] Are all changes manually tested?

- [ ] Does this change require any documentation changes?

- [ ] Are relevant unit and component test-cases added?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

## Summary by CodeRabbit

- **New Features**
	- Enhanced data handling with new utility functions for object and identifier validation, and unique value extraction.
	- Improved HubSpot search capabilities within the application.

- **Bug Fixes**
	- Updated existing data retrieval function for better performance and accuracy.

- **Tests**
	- Added test cases for new utility functions to ensure reliability and correctness.

### CODE REVIEW

**Comment by utsabc** [comment]
Can we add unit tests for the utils.js

**Comment by koladilip** [comment]
@rabbitai review

**Comment by koladilip** [comment]
@coderabbitai review

File: src/v0/destinations/hs/util.js [hunk 1]
```diff
@@ -533,7 +524,7 @@ const getExistingData = async (inputs, destination) => {
       Config.authorizationType === 'newPrivateAppApi'
         ? endpoint
         : `${endpoint}?hapikey=${Config.apiKey}`;
-    searchResponse =
+    const searchResponse =
```
Comment (koladilip) [inline src/v0/destinations/hs/util.js:527]: move: 
const endpoint = IDENTIFY_CRM_SEARCH_ALL_OBJECTS.replace(':objectType', objectType);
    const endpointPath = `objects/:objectType/search`;
    const url =
      Config.authorizationType === 'newPrivateAppApi'
        ? endpoint
        : `${endpoint}?hapikey=${Config.apiKey}`;
        
       these outside the while block

File: src/v0/destinations/hs/util.js [hunk 2]
```diff
@@ -545,30 +536,90 @@ const getExistingData = async (inputs, destination) => {
...
+
+  return searchResults;
+};
+
+/**
+ * DOC: https://developers.hubspot.com/docs/api/crm/search
+ * @param {*} inputs
+ * @param {*} destination
+ */
+const getExistingData = async (inputs, destination) => {
```
Comment (koladilip) [inline src/v0/destinations/hs/util.js:575]: which existing data, be specific

File: src/v0/destinations/hs/util.js [hunk 3]
```diff
@@ -464,61 +467,49 @@ const getEventAndPropertiesFromConfig = (message, destination, payload) => {
...
-  inputs.map(async (input) => {
+  return { objectType, identifierType };
+};
+
+/**
+ * Returns values for search api call
+ * @param {*} inputs
+ * @returns
+ */
+const extractUniqueValues = (inputs) => {
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js]: addressed

File: src/v0/destinations/hs/util.js [hunk 4]
```diff
@@ -464,61 +467,49 @@ const getEventAndPropertiesFromConfig = (message, destination, payload) => {
...
+/**
+ * Returns hubspot records
+ * @param {*} data
+ * @param {*} requestOptions
+ * @param {*} objectType
+ * @param {*} identifierType
+ * @param {*} Config
+ * @returns
+ */
+const performHubSpotSearch = async (rqdata, requestOptions, objectType, identifierType, Config) => {
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js]: addressed

File: src/v0/destinations/hs/util.js (hunk 1 above)
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js:535]: addressed

File: src/v0/destinations/hs/util.js (hunk 2 above)
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js]: addressed

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

File: src/v0/destinations/hs/util.js [hunk 5]
```diff
@@ -464,76 +467,72 @@ const getEventAndPropertiesFromConfig = (message, destination, payload) => {
...
-    const endpoint = IDENTIFY_CRM_SEARCH_ALL_OBJECTS.replace(':objectType', objectType);
-    const endpointPath = `objects/:objectType/search`;
-
-    const url =
-      Config.authorizationType === 'newPrivateAppApi'
-        ? endpoint
-        : `${endpoint}?hapikey=${Config.apiKey}`;
-    searchResponse =
+    const searchResponse =
       Config.authorizationType === 'newPrivateAppApi'
```
Comment (koladilip) [inline src/v0/destinations/hs/util.js:536]: const httpPOST = async (url, data, options, statTags = {})  has this signature but 
```
await httpPOST(url, requestData, {
            destType: 'hs',
            feature: 'transformation',
            endpointPath,
          });`

```
          code is wrong, should write like:
```
await httpPOST(url, requestData, {}, {
            destType: 'hs',
            feature: 'transformation',
            endpointPath,
          });
```

**Review by koladilip — COMMENTED** [review]
not full review

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

File: src/v0/destinations/hs/util.js (hunk 5 above)
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js:536]: addressed

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

File: src/v0/destinations/hs/util.js [hunk 6]
```diff
@@ -464,111 +467,172 @@ const getEventAndPropertiesFromConfig = (message, destination, payload) => {
...
```
Comment (koladilip) [inline src/v0/destinations/hs/util.js:536]: can we write like this:
```
requestOptions = Config.authorizationType === 'newPrivateAppApi' ? requestOptions : {};
await httpPOST(url, requestData, requestOptions, {
            destType: 'hs',
            feature: 'transformation',
            endpointPath,
          })
```

Comment (mihir-4116) [reply src/v0/destinations/hs/util.js]: addressed

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

File: src/v0/destinations/hs/util.js [hunk 7]
```diff
@@ -506,13 +506,7 @@ const extractIDsForSearchAPI = (inputs) => {
...
  * @returns
  */
-const performHubSpotSearch = async (
-  rqdata,
-  requestOptions,
-  objectType,
-  identifierType,
-  destination,
-) => {
+const performHubSpotSearch = async (rqdata, rqOptions, objectType, identifierType, destination) => {
```
Comment (koladilip) [inline src/v0/destinations/hs/util.js:509]: use reqData, reqOptions

Comment (mihir-4116) [reply src/v0/destinations/hs/util.js]: addressed

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Review by koladilip — DISMISSED** [review]

File: src/v0/destinations/hs/util.js [hunk 8]
```diff
@@ -464,111 +467,163 @@ const getEventAndPropertiesFromConfig = (message, destination, payload) => {
...
+/**
+ * Returns hubspot records
+ * @param {*} data
+ * @param {*} requestOptions
+ * @param {*} objectType
+ * @param {*} identifierType
+ * @param {*} destination
+ * @returns
+ */
+const performHubSpotSearch = async (
```
Comment (utsabc) [inline src/v0/destinations/hs/util.js:509]: Add documentation about the search api

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

File: src/v0/destinations/hs/util.js (hunk 8 above)
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js:510]: addressed

**Review by utsabc — DISMISSED** [review]

File: src/v0/destinations/hs/util.js [hunk 9]
```diff
@@ -464,111 +467,163 @@ const getEventAndPropertiesFromConfig = (message, destination, payload) => {
...
+ * Returns values for search api call
+ * @param {*} inputs
+ * @returns
+ */
+const extractIDsForSearchAPI = (inputs) => {
+  const values = inputs.map((input) => {
     const { message } = input;
     const { destinationExternalId } = getDestinationExternalIDInfoForRetl(message, DESTINATION);
-    values.push(destinationExternalId.toString().toLowerCase());
+    return destinationExternalId.toString().toLowerCase();
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js:494]: it's working

File: src/v0/destinations/hs/util.js [hunk 10]
```diff
@@ -464,111 +467,163 @@ const getEventAndPropertiesFromConfig = (message, destination, payload) => {
...
+      limit: SEARCH_LIMIT_VALUE,
+      after: 0,
+    };
+
+    const requestOptions = {
+      headers: {
+        'Content-Type': JSON_MIME_TYPE,
+        Authorization: `Bearer ${Config.accessToken}`,
+      },
+    };
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js:618]: TODO

File: src/v0/destinations/hs/util.js [hunk 11]
```diff
@@ -464,111 +467,164 @@ const getEventAndPropertiesFromConfig = (message, destination, payload) => {
...
+            },
+          ],
+        },
+      ],
+      properties: [identifierType],
+      limit: SEARCH_LIMIT_VALUE,
+      after: 0,
+    };
+
+    const requestOptions = {
```
Comment (koladilip) [inline src/v0/destinations/hs/util.js:613]: define this outside

Comment (mihir-4116) [reply src/v0/destinations/hs/util.js]: addressed

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Review by utsabc — APPROVED** [review]

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Review by koladilip — DISMISSED** [review]

**Review by koladilip — APPROVED** [review]

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/2907/test-report.html

**Review by koladilip — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**